cmd/podman/system_df.go

```diff
@@ -0,0 +1,363 @@
+	for _, usage := range diskUsages {
+		out = append(out, interface{}(usage))
+	}
+	return out
+}
+
+func getDiskUsage(ctx context.Context, runtime *libpod.Runtime) ([]systemDfDiskUsage, error) {
+	images, err := runtime.ImageRuntime().GetImages()
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to get images")
+	}
+	containers, err := runtime.GetAllContainers()
+	if err != nil {
+		return nil, errors.Wrapf(err, "error getting all containers")
+	}
+	volumes, err := runtime.GetAllVolumes()
+	if err != nil {
+		return nil, errors.Wrap(err, "error getting all volumes")
+	}
+	activeContainers, err := activeContainers(runtime)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error getting active containers")
+	}
+	imagesUsedbyCtrMap, imagesUsedbyActiveCtr, err := imagesUsedbyCtr(containers, activeContainers)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error getting getting images used by containers")
+	}
+	volumeUsedByContainerMap := volumeUsedByContainer(containers)
+
+	imageDiskUsage, err := getImageDiskUsage(ctx, images, imagesUsedbyCtrMap, imagesUsedbyActiveCtr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error getting disk usage of images")
+	}
+	containerDiskUsage, err := getContainerDiskUsage(containers, activeContainers)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error getting disk usage of containers")
+	}
+	volumeDiskUsage, err := getVolumeDiskUsage(volumes, volumeUsedByContainerMap)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error getting disk usage of volumess")
+	}
+	var systemDfDiskUsages []systemDfDiskUsage
+	systemDfDiskUsages = append(systemDfDiskUsages, imageDiskUsage)
+	systemDfDiskUsages = append(systemDfDiskUsages, containerDiskUsage)
+	systemDfDiskUsages = append(systemDfDiskUsages, volumeDiskUsage)
+	return systemDfDiskUsages, nil
+}
+
+func imageUniqueSize(ctx context.Context, images []*image.Image) (map[string]uint64, error) {
+	imgUniqueSizeMap := make(map[string]uint64)
+	for _, img := range images {
+		parentImg := img
+		for {
+			next, err := parentImg.GetParent()
+			if err != nil {
+				return nil, errors.Wrapf(err, "error getting parent of image %s", parentImg.ID())
+			}
+			if next == nil {
+				break
+			}
+			parentImg = next
+		}
+		imgSize, err := img.Size(ctx)
+		if err != nil {
+			return nil, err
+		}
+		if img.ID() == parentImg.ID() {
+			imgUniqueSizeMap[img.ID()] = *imgSize
+		} else {
+			parentImgSize, err := parentImg.Size(ctx)
+			if err != nil {
+				return nil, errors.Wrapf(err, "error getting size of parent image %s", parentImg.ID())
+			}
+			imgUniqueSizeMap[img.ID()] = *imgSize - *parentImgSize
+		}
+	}
+	return imgUniqueSizeMap, nil
+}
+
+func getImageDiskUsage(ctx context.Context, images []*image.Image, imageUsedbyCintainerMap map[string][]*libpod.Container, imageUsedbyActiveContainerMap map[string][]*libpod.Container) (systemDfDiskUsage, error) {
+	var (
+		numberOfImages       int
```

Comment:
numberofImages == len(images)

Reply:
not the same, middle layer images are not counted